Tests/LoggingTests/TestLogger.swift

```diff
@@ -31,7 +31,10 @@ internal struct TestLoggger: LogHandler {
     }
 
     func log(level: LogLevel, message: String, file: String, function: String, line: UInt) {
-        let metadata = self.metadata ?? MDC.global.metadata // use MDC unless set
+        // I don't think this would be a correct way of using those, rather a merge is needed:
+        // let metadata = self.metadata ?? MDC.global.metadata // use MDC unless set
+        let metadata = MDC.global.metadata.merging(self.metadata, uniquingKeysWith: { l, r in r })
```

Comment:
Whoops, just opened https://github.com/weissi/swift-server-logging-api-proposal/issues/7 but will consider your idea here in depth and reply in a moment :) Thanks for chiming in :)

Reply:
Thanks for clarifying Tom, it indeed seems to be about the different design assumptions we both operated on.

> here you solved it by a merge, but my design assumption was that merge is not always what you want. sometimes you'd want to be able and completely override the metadata. obviously we leave this to the LogHandler implementation, and the user can always do the merge at their level too.

Yeah, the merge with a global MDC perhaps is a not so fortunate application of this. I can agree that perhaps the local one may want to sometimes "win" etc... "it depends"™ is the killer key phrase here I guess. I agree that since this is only about the API really, we should keep it open enough to satisfy the use case as you lined it out as well, in case someone wants it.

For painting a fuller picture I want to explain a bit where the merging in my examples comes from: In applications I deal with the "hierarchy" is "natural" and always there, so lower levels are "more specific" one could say thus they may want to override, though I can see that if an application has no such layers perhaps it would care less.

Last example why the layers make most sense for me and some kinds of apps, layers being:
- "the context where you execute" which remains relatively stable, it contains some metadata which we may want to log,
- then per execution we have some "request dependent" metadata, the most typical examples would be anything that's a trace etc. this is still not really set by user code.
- user code may also set some metadata, on a logger we provided them, we do not want to remove these metadata, the users are in charge of clearing it.

So in this scheme "overriding" and "reset to the stable one" are the semantics I'm after.
The `Optional` does not block implementing this, but looked a bit weird to me, so if you have use cases which like keeping that optional information I'm happy to keep it in the API :-)

--- 

As we work on the general API here, I suppose keeping it open enough that both of our designs could work would be best, which in this case means keeping the `?` I suppose. I can live with that 😉 

TL;DR; Seems we keep the `?` to satisfy differentiating between "empty" and "did not set".
If an impl does not do not setting, it can always do [:] after all then.